// @wordpress/block-editor/build/components/rich-text/format-toolbar-container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _richText = require("@wordpress/rich-text");

var _blockControls = _interopRequireDefault(require("../block-controls"));

var _formatToolbar = _interopRequireDefault(require("./format-toolbar"));

var _store = require("../../store");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function InlineSelectionToolbar(_ref) {
  let {
    value,
    anchorRef,
    activeFormats
  } = _ref;
  const lastFormat = activeFormats[activeFormats.length - 1];
  const lastFormatType = lastFormat === null || lastFormat === void 0 ? void 0 : lastFormat.type;
  const settings = (0, _data.useSelect)(select => select(_richText.store).getFormatType(lastFormatType), [lastFormatType]);
  const selectionRef = (0, _richText.useAnchorRef)({
    ref: anchorRef,
    value,
    settings
  });
  return (0, _element.createElement)(InlineToolbar, {
    anchorRef: selectionRef
  });
}

function InlineToolbar(_ref2) {
  let {
    anchorRef
  } = _ref2;
  return (0, _element.createElement)(_components.Popover, {
    position: "top center",
    focusOnMount: false,
    anchorRef: anchorRef,
    className: "block-editor-rich-text__inline-format-toolbar",
    __unstableSlotName: "block-toolbar"
  }, (0, _element.createElement)("div", {
    className: "block-editor-rich-text__inline-format-toolbar-group"
  }, (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_formatToolbar.default, null))));
}

const FormatToolbarContainer = _ref3 => {
  let {
    inline,
    anchorRef,
    value
  } = _ref3;
  const hasInlineToolbar = (0, _data.useSelect)(select => select(_store.store).getSettings().hasInlineToolbar, []);

  if (inline) {
    return (0, _element.createElement)(InlineToolbar, {
      anchorRef: anchorRef
    });
  }

  if (hasInlineToolbar) {
    const activeFormats = (0, _richText.getActiveFormats)(value);

    if ((0, _richText.isCollapsed)(value) && !activeFormats.length) {
      return null;
    }

    return (0, _element.createElement)(InlineSelectionToolbar, {
      anchorRef: anchorRef,
      value: value,
      activeFormats: activeFormats
    });
  } // Render regular toolbar.


  return (0, _element.createElement)(_blockControls.default, {
    group: "inline"
  }, (0, _element.createElement)(_formatToolbar.default, null));
};

var _default = FormatToolbarContainer;
exports.default = _default;
